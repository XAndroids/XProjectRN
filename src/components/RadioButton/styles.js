import {StyleSheet} from "react-native";

export const styles = StyleSheet.create({
    radio: {
        justifyContent: 'center',
        alignItems: 'center',
        alignSelf: 'center'
    }
});
