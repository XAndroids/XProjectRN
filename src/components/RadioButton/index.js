import React from 'react'
import {TouchableOpacity, View} from 'react-native'
import PropTypes from 'prop-types'
import {styles} from './styles'

const DEFAULT_SIZE_MULTIPLIER = 0.7;
const DEFAULT_OUTER_BORDER_WIDTH_MULTIPLIER = 0.15;
/**
 * 单选按钮
 * 参考：https://github.com/mmazzarolo/react-native-radio-button/blob/master/RadioButton.js
 */
const RadioButton = (props) => {
    const {size, innerColor, outerColor, isSelected, onPress} = props;

    //外圈样式
    const outerStyle = {
        borderColor: outerColor,
        width: size + size * DEFAULT_SIZE_MULTIPLIER,
        height: size + size * DEFAULT_SIZE_MULTIPLIER,
        borderRadius: (size + size * DEFAULT_SIZE_MULTIPLIER) / 2,
        borderWidth: size * DEFAULT_OUTER_BORDER_WIDTH_MULTIPLIER
    };

    //内圈样式
    const innerStyle = {
        width: size,
        height: size,
        borderRadius: size / 2,
        backgroundColor: innerColor
    };

    return (
        <TouchableOpacity testID="radiobutton_touchableopacity_outer" style={[props.style, styles.radio, outerStyle]}
                          onPress={onPress}>
            {isSelected ? <View testID="radiobutton_view_inner" style={innerStyle}/> : null}
        </TouchableOpacity>
    )
};

RadioButton.propTypes = {
    size: PropTypes.number,
    innerColor: PropTypes.string,
    outerColor: PropTypes.string,
    isSelected: PropTypes.bool,
    onPress: PropTypes.func
};

RadioButton.defaultProps = {
    size: 16,
    innerColor: '#058b7d',
    outerColor: '#058b7d',
    isSelected: false,
    onPress: () => null
};

export default RadioButton

