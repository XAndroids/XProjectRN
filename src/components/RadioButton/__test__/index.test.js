import React from "react";
import {render, shallow} from "react-native-testing-library";
import RadioButton from "../index";

describe('Radio Button property test', () => {
    test('property default outer test', () => {
        //外部元素默认样式测试
        const {getByTestId, queryByTestId} = render(<RadioButton/>);
        const outComponent = getByTestId('radiobutton_touchableopacity_outer');

        //外部元素存在
        expect(getByTestId('radiobutton_touchableopacity_outer')).toBe(outComponent);
        //内部元素不存在
        //getByTestId如果没有元素，throw an error
        expect(() => getByTestId('radiobutton_view_inner')).toThrow('No instances found with props: {\"testID\":\"radiobutton_view_inner\"}');
        //queryByTestId如果没有元素，返回null
        expect(queryByTestId('radiobutton_view_inner')).toBeNull();

        //外部元素默认样式
        const outStyles = outComponent.props.style[2];
        expect(outStyles.borderColor).toBe('#058b7d');
        expect(outStyles.width).toBe(27.2);
        expect(outStyles.height).toBe(27.2);
        expect(outStyles.borderRadius).toBe(13.6);
        expect(outStyles.borderWidth).toBe(2.4);
    });

    test('property default inner test', () => {
        //内部元素默认样式测试
        const {getByTestId} = render(<RadioButton isSelected={true}/>);
        const innerComponent = getByTestId('radiobutton_view_inner');

        //内部元素存在
        expect(getByTestId('radiobutton_view_inner')).toBe(innerComponent);

        //内部元素默认样式
        const innerStyles = innerComponent.props.style;
        expect(innerStyles.width).toBe(16);
        expect(innerStyles.height).toBe(16);
        expect(innerStyles.borderRadius).toBe(8);
        expect(innerStyles.backgroundColor).toBe('#058b7d');
    });

    test('size property test', () => {
        //使用size和isSelected属性渲染RadioButton
        const {getByTestId} = render(<RadioButton size={3} isSelected={true}/>);
        const outComponent = getByTestId('radiobutton_touchableopacity_outer');
        const innerComponent = getByTestId('radiobutton_view_inner');

        //分别验证外部元素size相关样式
        const outStyles = outComponent.props.style[2];
        expect(outStyles.width).toBe(5.1);
        expect(outStyles.height).toBe(5.1);
        expect(outStyles.borderRadius).toBe(2.55);
        expect(outStyles.borderWidth).toBe(0.44999999999999996);

        //分别验证内部元素size相关样式
        const innerStyles = innerComponent.props.style;
        expect(innerStyles.width).toBe(3);
        expect(innerStyles.height).toBe(3);
        expect(innerStyles.borderRadius).toBe(1.5);
    });

    test('innerColor property test', () => {
        const {getByTestId} = render(<RadioButton innerColor={'#ffffff'} isSelected={true}/>);

        const innerComponent = getByTestId('radiobutton_view_inner');
        const innerStyles = innerComponent.props.style;
        expect(innerStyles.backgroundColor).toBe('#ffffff');
    });

    test('outColor property test', () => {
        const {getByTestId} = render(<RadioButton outerColor={'#131313'}/>);
        const outComponent = getByTestId('radiobutton_touchableopacity_outer');
        const outStyles = outComponent.props.style[2];
        expect(outStyles.borderColor).toBe('#131313');
    });

    test('isSelected property test', () => {
        //FIXME getByName不能使用？？
        const {getByTestId} = render(<RadioButton isSelected={true}/>);
        const innerComponent = getByTestId('radiobutton_view_inner');
        expect(getByTestId('radiobutton_view_inner')).toBe(innerComponent);

        const {queryByTestId} = render(<RadioButton isSelected={false}/>);
        expect(queryByTestId('radiobutton_view_inner')).toBeNull();
    });
});

describe('Radio Button fun test', () => {
    test('onPress fun test', () => {
        const fn = jest.fn();
        const {getByTestId} = render(<RadioButton onPress={fn}/>);
        const outComponent = getByTestId('radiobutton_touchableopacity_outer');
        outComponent.props.onPress();
        expect(fn).toHaveBeenCalled();
    });
});

describe('Radio Button Snapshot test', () => {
    //使用不同的属性配置，生成镜像，覆盖更多的代码
    test('default property Snapshot test', () => {
        const {toJSON} = render(<RadioButton/>);
        expect(toJSON()).toMatchSnapshot();
    });

    test('custom property Snapshot test', () => {
        const fn = jest.fn();
        const {toJSON} = render(<RadioButton isSelected={true} size={6} innerColor={'#aaaaaa'} outerColor={'#bbbbbb'}
                                             onPress={fn}/>);
        expect(toJSON()).toMatchSnapshot();
    });
});

describe('Radio Button Shallow test', () => {
    test('default property radio button Shallow snapshot test', () => {
        const fn = jest.fn();
        //shallow底层调用react-test-render/shallow
        const {output} = shallow(<RadioButton isSelected={true} size={6} innerColor={'#aaaaaa'} outerColor={'#bbbbbb'}
                                              onPress={fn}/>);
        //浅层渲染后，进行镜像测试
        expect(output).toMatchSnapshot();
    });

    test('inner component Shallow snapshot test', () => {
        const {getByTestId} = render(<RadioButton isSelected={true}/>);
        //指定"层"子组件进行浅层测试
        const {output} = shallow(getByTestId('radiobutton_view_inner'));
        expect(output).toMatchSnapshot();
    });

    test('default property radio button Shallow test', () => {
        const fn = jest.fn();
        const {output} = shallow(<RadioButton isSelected={true} size={6} innerColor={'#aaaaaa'} outerColor={'#bbbbbb'}
                                              onPress={fn}/>);
        //浅层渲染后，进行元素的expect
        expect(output.type.displayName).toBe('TouchableOpacity');
    });
});
