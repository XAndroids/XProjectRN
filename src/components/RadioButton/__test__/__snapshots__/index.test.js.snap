// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Button Shallow test default property radio button Shallow snapshot test 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "justifyContent": "center",
      },
      Object {
        "borderColor": "#bbbbbb",
        "borderRadius": 5.1,
        "borderWidth": 0.8999999999999999,
        "height": 10.2,
        "width": 10.2,
      },
    ]
  }
  testID="radiobutton_touchableopacity_outer"
>
  <Component
    style={
      Object {
        "backgroundColor": "#aaaaaa",
        "borderRadius": 3,
        "height": 6,
        "width": 6,
      }
    }
    testID="radiobutton_view_inner"
  />
</TouchableOpacity>
`;

exports[`Radio Button Shallow test inner component Shallow snapshot test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#058b7d",
      "borderRadius": 8,
      "height": 16,
      "width": 16,
    }
  }
  testID="radiobutton_view_inner"
/>
`;

exports[`Radio Button Snapshot test custom property Snapshot test 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "borderColor": "#bbbbbb",
      "borderRadius": 5.1,
      "borderWidth": 0.8999999999999999,
      "height": 10.2,
      "justifyContent": "center",
      "opacity": 1,
      "width": 10.2,
    }
  }
  testID="radiobutton_touchableopacity_outer"
>
  <View
    style={
      Object {
        "backgroundColor": "#aaaaaa",
        "borderRadius": 3,
        "height": 6,
        "width": 6,
      }
    }
    testID="radiobutton_view_inner"
  />
</View>
`;

exports[`Radio Button Snapshot test default property Snapshot test 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "borderColor": "#058b7d",
      "borderRadius": 13.6,
      "borderWidth": 2.4,
      "height": 27.2,
      "justifyContent": "center",
      "opacity": 1,
      "width": 27.2,
    }
  }
  testID="radiobutton_touchableopacity_outer"
/>
`;
